### PR TITLE
Implement yaml

### PR DESCRIPTION
**Pull request by JayOrten**
Note 1/27: I've tested all of these changes...except train and gridsearch fully, because I couldn't get a single gpu to test on. We can hold off until we've had a chance to fully test those scripts, but any needed changes would be minor.

This PR implements YAML configuration files, intended to be used project wide for any configuration needs.

The design choice of the YAML file is intended to eliminate repetition of commonly used parameters across code, as well as simplify future changes and refactors. The power of the YAML file is that it is very flexible: a developer can add to it as they see fit for whatever parameters they need, with the knowledge that the user will be aware of what parameters are needed when they run the code.

A few important notes about this design:
- The YAML is intended to store any path information. Absolute paths work best. The idea is that the YAML allows us to decouple any path information or logic from the code; therefore, when the user is executing the code, they can just look into the YAML to understand all of the paths needed/being used. This brings the flexibility that a custom YAML can work with any directory configuration, which is especially useful on the compute clusters. The downside is that the YAML may acquire multiple, somewhat repetitious paths. However, each path is clearly seen, and should serve a distinct purpose across multiple scripts.

- The YAML file eliminates repetition across code. The YAML is intended to deliver parameters straight to where they are used. That way, when code needs to be changed, only the YAML and the place where the parameter is used needs to be changed. The downside of this is that each script is only taking the config object as an argument, and using that everywhere. While being less transparent, this pattern is fairly common and allows more flexibility both during development and use.

- If there is a mistake in any parameter configuration, it is very easy to track it down in the code, straight to the YAML. If a parameter in the YAML is missing, it will be easily seen. This design also assumes no default arguments, which makes for a cleaner workflow, as any parameters being used are stored in the single YAML file.

The major contributions/changes of this PR are:

- Removes any argument parsing from major scripts, and replaces them with YAML -> config object logic. Adds a struct class to work with this
- Adds a configuration folder with default config, as well as a user folder to contain custom configs, to be ignored by git.
- Refactors/renames many common parameters and paths across scripts to be uniform
- Changes all scripts in scripts folder to be standard and using YAML. Adds an additional user_scripts folder for custom scripts, to be ignored by git.

### CODE REVIEW

File: configs/template_config.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+layers: 2
+# model_type (str): Name of model architecture to train.
+model_type: retnet # Choices: "retnet", "transformer"
+# heads (int): Number of heads. Head architecture changes based on model.
+heads: 4
+# seq_len (int): Sequence length (context window size).
+seq_len: 128
+# value_embed_dim (int): Value embed dimension size.
+value_embed_dim: 12
+# vocab_size (int): Maximum vocabulary size (number of unique tokens in vocabulary.
```
Comment (KimballNJardine) [inline configs/template_config.yaml]: Missing ")" at end of line.

File: configs/template_config.yaml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+tokenized_dataset_path: "/YOUR/PATH/HERE/data/tokenized_dataset/wikitext"
+
+# Train Model
+# Path to model folder
+models_path: "/YOUR/PATH/HERE/data/models"
+
+# Tensorboard path
+tboard_path: ~
+
+splits:
```
Comment (KimballNJardine) [inline configs/template_config.yaml:48]: An explanation of splits would be good: Example:
\# Splits
\# Fraction Train, Fraction Validation, Fraction Test

File: configs/template_config.yaml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+#        train a tokenizer, tokenize the data, train a model, and evaluate the model.
+#
+#        To prepare a YAML config for usage, create a copy of template_config.yaml 
+#        in the user_configs folder and fill in the necessary parameters.
+#        Path names need to be specified properly. These should be absolute paths.
+#        A suggestion for path structure is given in the template.
+#        The YAML config file is then passed as an argument to any scripts.
+#
+#        For example, to train a model, run:
+#        python train_model.py ./configs/my_config.yaml
```
Comment (KimballNJardine) [inline configs/template_config.yaml]: I am not certain, so correct me if I am wrong, but would it be a good idea to suggest the following instead?:

python train_model.py ./configs/user_configs/my_config.yaml.

If I understand the new gitignore file, that folder is where everyone should put their YAML files.

File: download_data.py [hunk 4]
```diff
@@ -19,22 +18,25 @@ def download_data(
...
+            name=config.dataset_subset,
             split="all",
             trust_remote_code=True)
 
     # check if dataset is of type datasets.arrow_dataset.Dataset
     if isinstance(dataset, datasets.arrow_dataset.Dataset):
-        filename = dataset_subset + ".parquet"
+        filename = config.dataset_subset + ".parquet"
+        dataset.to_parquet(dataset_dir / filename)
+        filename = config.dataset_subset + ".parquet"
```
Comment (KimballNJardine) [inline download_data.py]: These lines are duplicated. Please remove duplicate lines 37 and 38.

File: download_data.py [hunk 5]
```diff
@@ -44,16 +46,12 @@ def download_data(
...
+    config_path =args[1]
 
-    parser.add_argument("--dataset-name", type=str, required=True,
-        help="Hugging Face dataset name. Should also set --dataset-subset.")
-    parser.add_argument("--dataset-subset", type=str, required=True,
-        help="Subset/config to use for Hugging Face dataset.")
-    parser.add_argument("--datasets-dir", type=str, required=True,
-        help="Path to directory in which Hugging Face datasets are downloaded.")
+    with open(config_path, 'r') as f:
+        config = yaml.safe_load(f)
```
Comment (KimballNJardine) [inline download_data.py:51]: The white space here is excessive, in my opinion. This is repeated in the other files though, so it is consistent. Let's decide on if to keep all the blank lines or condense them before approving the pull request.

File: grid_search.py [hunk 6]
```diff
@@ -148,21 +129,12 @@ def grid_search(
...
-        help="Path to directory in which Hugging Face datasets are downloaded.")
-    parser.add_argument("-r", "--rand-seed", type=int, default=None,
-        help="Random seed to use, allowing more reproducible results.")
-    parser.add_argument("--tokenizer-folder", type= str, required=True,
-        help="Path to the file where the tokenizer will be saved")
-    parser.add_argument("--vocab-size", type=int, required=True,
-        help="Maximum number of unique tokens in vocabulary.")
+    args = sys.argv
+    config_path =args[1]
+
```
Comment (KimballNJardine) [inline grid_search.py:136]: See earlier comment about white space.

File: scripts/download_data.sh
Comment (KimballNJardine) [inline scripts/download_data.sh:1]: We may want to rename the scripts folder "generic_scripts", or something like that. That is not a requirement, but something to discuss.

File: tokenize_data.py [hunk 7]
```diff
@@ -1,112 +1,59 @@
...
-        required=True,
-        help="Name of tokenized data.")
-    parser.add_argument(
-        "--tokenizer-folder",
-        type=str,
-        required=True,
-        help="Folder to save tokenizer to.")
+    args = sys.argv
+    config_path =args[1]
+
```
Comment (KimballNJardine) [inline tokenize_data.py:52]: See white space comments from earlier.

File: train_model.py [hunk 8]
```diff
@@ -153,20 +99,20 @@ def train_model(
...
     model_label = f"{datetime.now().strftime('%Y-%m-%d-%H:%M:%S')}_" + \
-        f"{model_type}_{total_params}"
+        f"{config.model_type}_{total_params}"
 
     # Make sure dataset is pre-downloaded
-    dataset_dir = Path(datasets_dir) / dataset_name
-    assert dataset_dir.exists(), \
-        f"The directory with data, {dataset_dir}, doesn't exist!"
-    print(f"\nUsing dataset directory {dataset_dir}")
+    # dataset_dir = Path(datasets_dir) / dataset_name
```
Comment (KimballNJardine) [inline train_model.py]: Decide if we want to remove the code entirely, or leave it as comments.

File: train_model.py [hunk 9]
```diff
@@ -176,73 +122,73 @@ def train_model(
     print(f"Saving weight files in {weights_dir}")
 
     # Initialize tokenizers directory
-    tokenizers_dir = Path(data_dir) / "tokenizers"
-    tokenizers_dir.mkdir(parents=False, exist_ok=True)
-    print(f"Saving tokenizer files in {tokenizers_dir}")
+    # tokenizers_dir = Path(models_path) / "tokenizers"
+    # tokenizers_dir.mkdir(parents=False, exist_ok=True)
+    # print(f"Saving tokenizer files in {tokenizers_dir}")
```
Comment (KimballNJardine) [inline train_model.py]: Decide if we want to remove code entirely or leave as comments.

File: train_model.py [hunk 10]
```diff
@@ -176,73 +122,73 @@ def train_model(
...
     else:
-        tboard_log_dir = f"{tboard_dir}/logs/{model_label}"
+        tboard_log_dir = f"{config.tboard_path}/logs/{model_label}"
     writer = SummaryWriter(log_dir=tboard_log_dir)
     print(f"Saving TensorBoard logs in {tboard_log_dir}")
 
     # Save all the variables in args as JSON inside folder
     json_string = json.dump(
-        obj=arg_dict,
+        obj=arg_table,
```
Comment (KimballNJardine) [inline train_model.py]: Double check if arg_dict is used for anything else. If no, remove arg_dict (created at start of method.)

File: train_model.py [hunk 11]
```diff
@@ -390,59 +336,11 @@ def train_model(
...
-    parser.add_argument("--value-embed-dim", type=int, default=1280,
-        help="Value embed dimension size.")
-    parser.add_argument("--vocab-size", type=int, required=True,
-        help="Maximum number of unique tokens in vocabulary.")
-
-    args = parser.parse_args()
-    train_model(**vars(args))
+    args = sys.argv
+    config_path =args[1]
+
```
Comment (KimballNJardine) [inline train_model.py:332]: See earlier comment about white space.

File: train_tokenizer.py [hunk 12]
```diff
@@ -102,62 +95,18 @@ def train_tokenizer(
...
-        "--vocab-size",
-        type=int,
-        required=True,
-        help="Maximum vocabulary size.")
-
-    args = parser.parse_args()
-    train_tokenizer(**vars(args))
+    args = sys.argv
+    config_path =args[1]
+
```
Comment (KimballNJardine) [inline train_tokenizer.py:107]: See earlier comments about white space.

**Review by KimballNJardine — CHANGES_REQUESTED** [review]
This really cleans up the code. I have some initial thoughts and issues that stuck out to me, but this looks really good.

File: configs/template_config.yaml (hunk 3 above)
Comment (JayOrten) [reply configs/template_config.yaml]: Good catch

File: train_model.py (hunk 10 above)
Comment (JayOrten) [reply train_model.py]: It's used up above for something else.

File: grid_search.py (hunk 6 above)
Comment (JayOrten) [reply grid_search.py:136]: discussed

File: tokenize_data.py (hunk 7 above)
Comment (JayOrten) [reply tokenize_data.py:52]: discussed

File: scripts/download_data.sh
Comment (JayOrten) [reply scripts/download_data.sh:1]: discussed

**Review by nprisbrey — APPROVED** [review]
I added some tweaks in the latest commits. As always, feel free to push back against them if anything doesn't seem correct. Now waiting for @KimballNJardine's approval.

File: scripts/train_tokenizer.sh
Comment (KimballNJardine) [inline scripts/train_tokenizer.sh:1]: Note comment for scripts/download_data.sh below.

File: scripts/download_data.sh
Comment (KimballNJardine) [inline scripts/download_data.sh:1]: For all the base scripts it does the following:
```python3 ../{name_of_shell_file}.py ../configs/user_configs/YOUR_CONFIG_HERE.yaml```.

This file structure implicitly suggests that it is being called while the user's working directory is ```/scripts```.
However, user scripts should be stored in ```/scripts/user_scripts```.
Going up one level would be in ```/scripts```, not the top level directory, so neither ```../{name_of_shell_file}.py``` nor  ```../configs/user_configs/YOUR_CONFIG_HERE.yaml``` could be found.

**I recommend changing all sh files to the following:
```python3 ../../{name_of_shell_file}.py ../../configs/user_configs/YOUR_CONFIG_HERE.yaml```
(Note the extra ```../``` on each part of the sh script)**

If I have misunderstood the file directories, please let me know. I am also willing to drop the point if everyone envisions calling the script another way, like ```sh user_scripts/download_data.sh``` while cwd is ```/scripts```.

File: scripts/grid_search.sh
Comment (KimballNJardine) [inline scripts/grid_search.sh:1]: Note comment for scripts/download_data.sh above.

File: scripts/tokenize_data.sh
Comment (KimballNJardine) [inline scripts/tokenize_data.sh:1]: Note comment for scripts/download_data.sh above.

File: scripts/train_model.sh
Comment (KimballNJardine) [inline scripts/train_model.sh:1]: Note comment for scripts/download_data.sh above.

**Review by KimballNJardine — CHANGES_REQUESTED** [review]
I have a minor request regarding the .sh files. It does not need to be changed per say, but it needs to be discussed. Otherwise, good.

File: scripts/download_data.sh
Comment (JayOrten) [reply scripts/download_data.sh:1]: Good point, I agree. I will make those changes

File: scripts/train_tokenizer.sh
Comment (JayOrten) [reply scripts/train_tokenizer.sh:1]: changed

File: scripts/download_data.sh
Comment (JayOrten) [reply scripts/download_data.sh:1]: changed

**Review by KimballNJardine — APPROVED** [review]
Looks good to me.

**Pull request merged**